### DISCUSSION TITLE
Embedded file in a docstring

### DISCUSSION THREAD

**Discussion by gadsc** [mkdocstrings/python#122] [Q&A] [answered]
Hello! Hope you are doing good :)
Do you know if it's possible to reference a code snippet inside the docstring? Something [like this](https://github.com/stevemar/code-reference-in-readme/tree/main). I've tried many approaches without any success :(

Btw awesome library 🎉 

Thanks in advance.

**Comment by pawamoy** [comment]
That's a good idea for a new Markdown extension I guess :slightly_smiling_face:

**Reply by gadsc** [reply]
Hello, thanks for the quick answer!
Do you mean a [mkdocstring extension](https://mkdocstrings.github.io/python/usage/extensions/), [handler extension](https://mkdocstrings.github.io/usage/handlers/#handler-extensions) or other kind of extension?
I was able to render the code inside a markdown file creating a pluglet for mkdoc directly, but when I try to use it in the docstring it's just escaping my function 😭 

Another question not related to it, do you know if it's possible to render the full enum path value? Right now when I have a enum it's just printing `value`, example:
```python
FORMATS = [ value, value, value]
```

The real code is something like:
```python
class Format(Enum):
    JDBC = "jdbc",
    [...]

FORMATS = [ Format.JDBC.value, [...]]
```

**Reply by pawamoy** [reply]
I mean a [Python-Markdown extension](https://python-markdown.github.io/extensions/api/). Basically an inline processor that matches against GitHub links, fetches the specified code lines through GitHub's API, and displays them like on GitHub. Or a different kind of processor which handles different markup than simple GitHub links.

An alternative would be a custom [pymdownx superfence](https://facelessuser.github.io/pymdown-extensions/extensions/superfences/). It's easier to develop, but would be more verbose for users:

````
```github-snippet
your github link
```
````

> I was able to render the code inside a markdown file creating a pluglet for mkdoc directly, but when I try to use it in the docstring it's just escaping my function 😭

Other pluglets (I suppose through mkdocs-macros?) or plugins can only see the docstrings once they're converted to HTML. They cannot see the original docstrings written in Markdown (unless hacks with mkdocs-macros). If you share your code, and the inputs/outputs you get, I can try to provide more guidance.

---

Regarding enum values, can you open a new issue? I've experienced the same thing :+1:

**Reply by gadsc** [reply]
I tried through mkdocs-macros, [based on this comment](https://github.com/mkdocs/mkdocs/issues/692#issuecomment-427056366), only changing that right now we should use `env.macro` as decorator instead `macro`.
This is how I tried to use it:
```python
class ExecEnv(object):
    """Represents the basic resources regarding the engine execution environment.

    Currently, it is solely used to encapsulate the logic to get a Spark session.
    ```
        {{ external_markdown("docs/test.md") }}
    ```

    Attributes:
        SESSION: Spark Session

    Examples:
        {{ external_markdown("docs/test.md") }}
    """
    [...]
```

This was the output:
![image](https://github.com/mkdocstrings/python/assets/3172171/18e11eeb-98c3-41d4-82f8-c3311482be34)

Going to test the py-markdown and the superfence and post an answer here, thanks a lot for the answer :)

I've created the [enum issue](https://github.com/mkdocstrings/python/issues/124)

**Reply by pawamoy** [reply, marked as answer]
Take a look at this: https://github.com/mkdocstrings/mkdocstrings/issues/615#issuecomment-1725838201.
It should help you getting macros to work in docstrings :slightly_smiling_face:

**Reply by gadsc** [reply]
It worked perfectly :)
Thanks for this magic trick :D
![magic](https://media2.giphy.com/media/v1.Y2lkPTc5MGI3NjExb2ZhMmJ4NG55bGZxMGhwdXA5anl5aDlwcjFhbDl6d3c2OTNkZzNjcSZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/x5oW1NcJJYQi87lmg3/giphy.gif)
